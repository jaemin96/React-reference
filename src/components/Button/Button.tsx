import { ReactNode } from 'react';

interface ButtonProps {
  primary?: boolean;
  backgroundColor?: string;
  size?: 'medium' | 'small' | 'large';
  children: ReactNode;
  type?: 'submit' | 'button' | 'reset';
  onClick?: () => void;
}

export const Button = ({
  primary = false,
  size = 'medium',
  backgroundColor,
  type = 'submit',
  onClick,
  ...props
}: ButtonProps) => {
  const mode = primary ? 'button--primary' : 'button--secondary';

  return (
    <button
      type={type}
      className={['button', `button--${size} ${mode}`].join(' ')}
      style={{ backgroundColor }}
      onClick={onClick}
      {...props}
    >
      {props.children}
    </button>
  );
};
