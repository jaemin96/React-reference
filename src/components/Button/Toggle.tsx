import React from 'react';

interface ToggleProps {
  isToggled?: boolean;
}

export const Toggle = ({ isToggled = false }: ToggleProps) => {
  const [visible, setVisible] = React.useState(isToggled);
  const onOff = visible ? 'toggle--on' : 'toggle--off';

  const handleVisible = () => {
    setVisible(!visible);
  };

  return (
    <div className={['toggle', `${onOff}`].join(' ')} onClick={handleVisible} data-testid="toggle">
      <div className="toggle--circle"></div>
    </div>
  );
};
