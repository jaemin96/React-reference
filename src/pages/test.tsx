import React, { ReactElement } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Button } from '@mui/material';
import { RootState } from '../redux/modules';
import { QuestionForm, Spinner } from '../components';
import { Increase, Decrease } from '../redux/modules/counter';

const TestPage = (): ReactElement => {
  const dispatch = useDispatch();
  const { num } = useSelector((state: RootState) => state.counter);

  return (
    <div>
      <div>
        <h1>State management</h1>
        <span>current value : {num}</span>
      </div>
      <Button className="increase-btn" onClick={() => dispatch(Increase())}>
        +
      </Button>
      <Button className="decrease-btn" onClick={() => dispatch(Decrease())}>
        -
      </Button>
      <div>
        <QuestionForm />
      </div>
      <div>
        <Spinner />
      </div>
    </div>
  );
};

export default TestPage;
