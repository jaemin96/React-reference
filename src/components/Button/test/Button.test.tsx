import { Button } from '../Button';
import { render, screen } from '@testing-library/react';
import { userEvent } from '@storybook/test';

describe('Button', () => {
  test('render button component with children', () => {
    render(<Button size={'medium'}>Click</Button>);
    const buttonElement = screen.getByText(/Click/i);
    expect(buttonElement).toBeInTheDocument();
  });

  test('button correct classes based on props', () => {
    const { container } = render(
      <Button primary size={'large'}>
        Click
      </Button>
    );
    const buttonElement = container.firstChild;
    expect(buttonElement).toHaveClass('button--primary'); // mode check
    expect(buttonElement).toHaveClass('button--large'); // size check
  });

  test('handle onClick event', async () => {
    const handleClick = jest.fn(); // jest function mocking
    render(
      <Button size={'large'} onClick={() => handleClick()}>
        Click
      </Button>
    );
    const buttonElement = screen.getByText(/Click/i);
    await userEvent.click(buttonElement);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test('snapshot', () => {
    const { asFragment } = render(
      <Button primary size={'small'}>
        Click
      </Button>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
