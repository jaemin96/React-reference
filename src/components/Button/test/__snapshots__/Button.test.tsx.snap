// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button snapshot 1`] = `
<DocumentFragment>
  <button
    class="button button--small button--primary"
    type="submit"
  >
    Click
  </button>
</DocumentFragment>
`;
